File: Leetcode_JS/longest-zigzag-binarytree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
var longestZigZag = function(root) {
    // We need to check every node for the longest zig zag path
    let count = 0


    function dfs(node, last, length = 0) {
        if (!node) return

        count = Math.max(count, length)
        dfs(node.left, "l", last !== "l" ? length + 1 : 1)
        dfs(node.right, "r", last !== "r" ? length + 1 : 1)
    }
    dfs(root, "r", 0)
    return count
};

File: Leetcode_Python/longest-zigzag-binarytree.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def longestZigZag(self, root):
        """
        :type root: Optional[TreeNode]
        :rtype: int
        """
        

        self.count = 0
        def dfs(node, left, right):
            if not node:
                return
            
            self.count = max(self.count, left, right)
            if node.left:
                dfs(node.left, right+1, 0)
            if node.right:
                dfs(node.right, 0, left+1)
        
        dfs(root, 0, 0)
        return self.count
                